### PR TITLE
Improve mutant rollback (Fix #236)

### LINKED ISSUE #236: Stryker cannot recover when unable to rollback a mutant
When Stryker fails to rollback a compilation error, the mutated assembly cannot be built and all tests result in survivors. This is obviously not a desirable behaviour.

Having a failsafe fallback strategy would prevent this for happening. For example, restoring the source file to its non mutated state would ensure the file can be compiled successfully.

### PR DESCRIPTION
**Pull request by dupdob**
Implements a backup rollback strategy when Stryker can't identify the mutation(s) causing compilation error(s): removes all mutants in the same method.
Error message is kept help prevent such occurence

### CODE REVIEW

File: src/Stryker.CLI/Stryker.CLI/CLIOptions.cs [hunk 1]
```diff
@@ -37,6 +37,15 @@ public static class CLIOptions
             JsonKey = "log-file"
         };
 
+        public static readonly CLIOption<bool> DevMode = new CLIOption<bool>
+        {
+            ArgumentName = "--dev-mode",
+            ArgumentShortName = "-dm <enable>",
+            ArgumentDescription = "Set to true to activate dev mode to ease debugging of Stryker.Net| Options [true, false (default)]",
+            DefaultValue = false,
+            JsonKey = "safe-mode"
```
Comment (Mobrockers) [inline src/Stryker.CLI/Stryker.CLI/CLIOptions.cs]: Is it called dev-mode or safe mode? Both are used currently. For example the unit test also uses --safe-mode

File: src/Stryker.CLI/Stryker.CLI/StrykerCLI.cs [hunk 2]
```diff
@@ -39,6 +39,7 @@ public int Run(string[] args)
             var configFilePathParam = CreateOption(app, CLIOptions.ConfigFilePath);
             var reporterParam = CreateOption(app, CLIOptions.Reporter);
             var logConsoleParam = CreateOption(app, CLIOptions.LogLevel);
+            var safeModeParam = CreateOption(app, CLIOptions.DevMode);
```
Comment (Mobrockers) [inline src/Stryker.CLI/Stryker.CLI/StrykerCLI.cs]: Dev mode or safe mode?

**Comment by dupdob** [comment]
Sorry for the PR spam, I want to fix issues I find vexing/blocking. I know it implies extra review effort, so thank you for that.
I am also driven by interest for mutation testing as well as sheer curiosity!

File: src/Stryker.CLI/Stryker.CLI/StrykerCLI.cs (hunk 2 above)
Comment (dupdob) [reply src/Stryker.CLI/Stryker.CLI/StrykerCLI.cs]: see previous: dev-mode

File: src/Stryker.CLI/Stryker.CLI/CLIOptions.cs (hunk 1 above)
Comment (dupdob) [reply src/Stryker.CLI/Stryker.CLI/CLIOptions.cs]: actually, it's dev-mode. Safe mode came from previous conversations.  need to fix that.

File: src/Stryker.Core/Stryker.Core/Compiling/CompilingProcess.cs [hunk 3]
```diff
@@ -59,7 +62,7 @@ public CompilingProcessResult Compile(IEnumerable<SyntaxTree> syntaxTrees, Memor
             {
                 LogEmitResult(emitResult);
                 // remove broken mutations
-                rollbackProcessResult = _rollbackProcess.Start(compiler, emitResult.Diagnostics);
+                rollbackProcessResult = _rollbackProcess.Start(compiler, emitResult.Diagnostics, false);
```
[the remark this answers is not in the document]
Comment (dupdob) [reply src/Stryker.Core/Stryker.Core/Compiling/CompilingProcess.cs]: I am fixing that.

**Comment by Mobrockers** [comment]
Don't worry, any effort to improve stryker and advance mutation testing is welcome here! We are quite happy with your contributions 👍

**Comment by dupdob** [comment]
I have no work pending on this PR, unless a new review raises concern

File: src/Stryker.CLI/Stryker.CLI.UnitTest/StrykerCLITests.cs [hunk 4]
```diff
@@ -159,15 +159,31 @@ public void StrykerCLI_WithLogLevelFileArgument_ShouldPassLogFileArgumentsToStry
...
 
-            mock.Verify(x => x.RunMutationTest(It.Is<StrykerOptions>(o => o.LogOptions.LogToFile == true)));
+            mock.Verify(x => x.RunMutationTest(It.Is<StrykerOptions>(o => o.LogOptions.LogToFile)));
+        }
+
+        [Theory]
+        [InlineData("--safe-mode")]
+        public void StrykerCLI_WithSafeModeArgument_ShouldPassSafeModeArgumentsToStryker(string argName)
+        {
+            StrykerOptions options = new StrykerOptions("", "Console", "", 1000, null, "trace", false, false, 1, 90, 80, 70);
```
[the remark this answers is not in the document]
Comment (Mobrockers) [reply src/Stryker.CLI/Stryker.CLI.UnitTest/StrykerCLITests.cs]: I misinterpreted the test, my bad. The test name still says safe mode though.

**Comment by Mobrockers** [comment]
I would like if you could pass the dev-mode flag to the integration test in the build pipeline. You can do so as follows.

Add this:

`arguments: '--dev-mode "true"'`
Beneath the following lines:

https://github.com/stryker-mutator/stryker-net/blob/69a05604bc989e3f3aed9d61473d45143acb682a/azure-pipelines.yml#L105

https://github.com/stryker-mutator/stryker-net/blob/69a05604bc989e3f3aed9d61473d45143acb682a/azure-pipelines.yml#L145

https://github.com/stryker-mutator/stryker-net/blob/69a05604bc989e3f3aed9d61473d45143acb682a/azure-pipelines.yml#L185

File: src/Stryker.CLI/Stryker.CLI.UnitTest/StrykerCLITests.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (dupdob) [reply src/Stryker.CLI/Stryker.CLI.UnitTest/StrykerCLITests.cs]: thanks for spotting this. I did a full solution search, but for _safe-mode_, did not occur to me to search for _searchmode_ as well.

**Comment by dupdob** [comment]
Done

**Pull request merged**